### ISSUE TITLE
NNBD: wrong inferred return type with `Iterable.map`

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/sdk#39701]
In the example program below the second variable is inferred as the type `String?` when afaik it should just be `String` ([nnbd dartpad link](https://dart-pad-nnbd.appspot.com/?id=8860898e79410eb3ee19e526f8ae50a4))

```dart
main() {
  var ints = [1,2,3];

  //  Inferred type is Iterable<String?>
  var strings = ints.map((o) => '$o');
  // Fails because its a List<String?> which isn't a List<String>
  List<String> stringList = strings.toList();

  // Interestingly this works, not sure why
  Iterable<String> explicitStrings = ints.map((o) => '$o');
  List<String> explicitStringList = explicitStrings.toList();
}
```

**Comment by srawlins** [comment]
When you say "Fails", what do you mean? You tagged this bug against the analyzer.

**Comment by jakemac53** [comment]
You get a static error because a `List<String?>` isn't a `List<String>`.

What seems like a bug is that the inference isn't what I would expect which is why I filed it on the analyzer. It might be working as intended - and if so we can close this, but its pretty undesirable behavior imo for a very common pattern. We might want to open a language issue in that case.

**Comment by jakemac53** [comment]
cc @leafpetersen @munificent

**Comment by jakemac53** [comment]
I talked with Bob about this and he agreed it looks like a bug in the inference.

Specifically it looks related to `=>` syntax - if you change the lambda to be `(o) { return '$o'; }` then it works as expected.

**Comment by munificent** [comment]
Right, for some reason, analyzer is inferring `(o) => '$o'` to have return type `String?` instead of `String`.

**Comment by leafpetersen** [comment]
cc @scheglov I suspect what's happening is that the analyzer is adding in `Null` to the inferred type, taking the upper bound of Null and String, and ending up with `String?`.  Just a guess though.

The relevant part of the proposed spec: https://github.com/dart-lang/language/pull/218/files#diff-8799cf34fe82cd1bb7ba3814a5e7fca4R236

**Comment by scheglov** [comment]
I have not read the revised type inference spec, but this issue was fixed (in some way) in 7707c2a84c85aadb08ff625d4c30486b5f078bd2.

**Comment by scheglov** [comment]
The issue has been fixed.
